src/plasma/plasma_manager.cc

```diff
@@ -642,19 +647,16 @@ int read_object_chunk(ClientConnection *conn, PlasmaRequestBuffer *buf) {
   }
   r = read(conn->fd, buf->data + conn->cursor, s);
 
-  if (r == -1) {
-    LOG_ERROR("read error");
-  } else if (r == 0) {
-    LOG_DEBUG("end of file");
+  if (r <= 0) {
+    LOG_ERROR("Read error");
+    return errno;
   } else {
     conn->cursor += r;
-  }
-  /* If the cursor is equal to the full object size, reset the cursor and we're
-   * done. */
-  if (conn->cursor == buf->data_size + buf->metadata_size) {
-    conn->cursor = 0;
-    return 1;
-  } else {
+    /* If the cursor is equal to the full object size, reset the cursor and
+     * we're done. */
+    if (conn->cursor == buf->data_size + buf->metadata_size) {
+      ClientConnection_finish_request(conn);
+    }
     return 0;
```

Comment:
The prior code looks like it was just returning `0` (meaning "not done with the transfer") in the event of an error. This makes it seem like it could just get into a situation where it kept trying to process the same object but never finished and so a connection between two managers would get blocked in some sense. Was that happening? I'm just wondering how this could possibly have been working if we weren't propagating errors from the call to `read` back to the caller..

Reply:
Yeah, this was the bug that we uncovered when testing with a live cluster. I don't think we run into it normally because the chance of a disconnection was small.